File: tday-web/tests/guardrails/task-row-first-line-alignment.test.ts
import { readFileSync } from "node:fs";
import { resolve } from "node:path";
import { describe, expect, it } from "vitest";

/**
 * THE CHECK CIRCLE BELONGS ON THE TITLE'S FIRST LINE, ON ALL THREE CLIENTS
 *
 * The report was one sentence — "make sure everywhere that checkbox bullet point is
 * inline with first line of task, it doesn't need to be centered between all the
 * lines" — and the word doing the work in it is "everywhere". Every task feed in the
 * app draws the same shape: a round toggle, a title that may wrap, and a list or
 * priority mark at the far end. Every one of them stacked that shape centred, which
 * is indistinguishable from correct while the title is one line and becomes the
 * defect the moment it is two: the toggle settles on the text column's MIDDLE, which
 * on a wrapped title is the gap BETWEEN line one and line two. The mark that is
 * supposed to be the line's bullet floats beside nothing.
 *
 * Eleven rows across three clients had it. A fix to the screen in the screenshot
 * would have left it in ten. That is what this file is for.
 *
 * ## Why a source read and not a render assertion
 *
 * jsdom lays nothing out — every box it reports is 0×0 — so the one thing that could
 * actually prove this, a measured offset, is exactly what the web test environment
 * cannot produce, and there is no emulator or simulator in the gate for the other
 * two. What CAN be checked is the thing that decides the layout: which alignment
 * each row asks for, and whether the box it puts its control in is the title's line
 * box or an accident. The arithmetic itself is asserted where the arithmetic lives —
 * `TaskRowFirstLineAlignmentTest` on Android sweeps the derivation across seven font
 * scales, and `TdayTaskRowSkeletonTests` pins iOS's baseline nudge to one value.
 *
 * ## The three idioms, and why they are three
 *
 * They are not a divergence, they are each client's existing answer to the same
 * question, and every one of them was already in the tree before this change:
 *
 *   - **iOS** aligns the `HStack` on `.firstTextBaseline` and hands the toggle an
 *     `.alignmentGuide` returning its own centre plus a nudge, because a `Button`
 *     holds no text and so reports no text baseline. `TodoListScreen` has done this
 *     since the timeline learned to wrap; the other three rows now do it too.
 *   - **Web** gives the control a box that is the title's line box (`h-5`, which is
 *     `leading-5`) and centres it there, with the row's content stacked `items-start`.
 *     `TaskRowSkeleton` was already built this way, and is why `leading-5` is the
 *     number both halves read.
 *   - **Android** cannot express either directly, so it derives: `Alignment.Top` plus
 *     an inset computed from the title style's own `lineHeight` and the control's
 *     height. At `fontScale = 1` that derivation returns exactly the 12 dp
 *     `TodoListScreen` had hand-written, which is the evidence it is the existing
 *     decision rather than a new one.
 *
 * Bare `items-start` / `Alignment.Top` is NOT the fix and is the trap worth naming:
 * top-aligning a 48 dp control against a 24 sp line box puts the glyph 12 dp low,
 * and because one is dp and the other sp the error moves with the user's font scale.
 */
describe("every task row hangs its control off the title's first line", () => {
  /** Paths under `tday-web/`. */
  const read = (path: string) =>
    readFileSync(resolve(__dirname, "../..", path), "utf8");

  /** Paths under the repository root — this claim reaches all three clients. */
  const readRepo = (path: string) =>
    readFileSync(resolve(__dirname, "../../..", path), "utf8");

  /**
   * Each web row, with the class strings that decide where its control lands.
   *
   * `h-5` is the title's `leading-5` line box written as a height — that pairing is
   * the whole technique, so both halves are asserted on every row. A row that kept
   * the `h-5` box and let its title drift to `leading-6` would be aligning to a line
   * it no longer draws.
   */
  const webRows = [
    "src/components/todo/component/TodoItemContainer.tsx",
    "src/features/floater/component/FloaterItemContainer.tsx",
    "src/features/calendar/component/CalendarClient.tsx",
    "src/features/completed/component/ItemContainer.tsx",
    "src/features/completed/component/CompletedFloaterItemContainer.tsx",
  ];

  it.each(webRows)("%s puts its control in the title's own line box", (path) => {
    const source = read(path);
    expect(source).toContain('className="flex h-5 shrink-0 items-center"');
    expect(source).toContain("leading-5");
  });

  it.each(webRows)("%s stacks its row content from the top", (path) => {
    const source = read(path);
    expect(source).toMatch(/items-start/);
  });

  /**
   * The trailing mark moves with the leading one, or the fix is half done.
   *
   * The screenshot had the priority flag floating in the same gap as the check
   * circle, and the two are the same claim: both are annotations ON the title, so
   * both read with its first line. The desktop hover toolbar deliberately does NOT
   * move — it is a menu for the whole task rather than a mark on its title, it is
   * 28 px against a 20 px line, and it stays centred on the row by being positioned
   * against a `self-stretch` box. That exemption is what makes `self-stretch` part of
   * the claim rather than decoration.
   */
  it.each([
    "src/components/todo/component/TodoItemContainer.tsx",
    "src/features/floater/component/FloaterItemContainer.tsx",
    "src/features/calendar/component/CalendarClient.tsx",
  ])("%s keeps its hover toolbar centred while its meta moves up", (path) => {
    const source = read(path);
    expect(source).toContain("self-stretch");
    expect(source).toContain('"flex h-5 items-center gap-2 transition-opacity"');
    // The toolbar still centres itself against that stretched box.
    expect(source).toContain("top-1/2 hidden -translate-y-1/2");
  });

  /**
   * The two Completed rows are the same row twice and carry the mark inline rather
   * than in a fading meta cluster, so they get their own spelling of the same claim.
   */
  it.each([
    ["src/features/completed/component/ItemContainer.tsx", "gap-2"],
    ["src/features/completed/component/CompletedFloaterItemContainer.tsx", "gap-1.5"],
  ])("%s puts its list mark on the first line", (path, gap) => {
    expect(read(path)).toContain(`className="flex h-5 shrink-0 items-center ${gap} pr-1"`);
  });

  const iosRows = [
    "ios-swiftUI/Tday/Feature/Todos/TodoListScreen.swift",
    "ios-swiftUI/Tday/Feature/Completed/CompletedScreen.swift",
    "ios-swiftUI/Tday/Feature/Calendar/CalendarScreen.swift",
    "ios-swiftUI/Tday/Feature/ScheduledTaskHome/ScheduledTaskHomeScreen.swift",
  ];

  it.each(iosRows)("%s aligns its task row on the first text baseline", (path) => {
    const source = readRepo(path);
    expect(source).toContain("HStack(alignment: .firstTextBaseline");
    // A Button reports no text baseline, so the guide is not optional decoration:
    // without it SwiftUI falls back to the button's BOTTOM edge and stands it low.
    expect(source).toMatch(
      /\.alignmentGuide\(\.firstTextBaseline\) \{ dimension in\n\s+dimension\[VerticalAlignment\.center\] \+ \w+\.(minimalRowBaselineNudge|checkBaselineNudge)/,
    );
  });

  /**
   * Android's derivation exists and is what the rows read.
   *
   * The Kotlin side has its own test for the arithmetic and for the set of files that
   * call this; what is added here is the cross-client claim — that the Android answer
   * to the same question is a DERIVATION rather than another hand-written number,
   * which is the only one of the three idioms that survives a font-scale change on
   * its own.
   */
  it("Android derives the inset from the title's line height rather than naming it", () => {
    const helper = readRepo(
      "android-compose/app/src/main/java/com/ohmz/tday/compose/core/ui/TdayTaskRowSkeleton.kt",
    );
    expect(helper).toContain("fun taskRowFirstLineAlignment(");
    expect(helper).toContain("spDimension(density, titleStyle.lineHeight, FallbackLineHeight)");

    const androidRows = [
      "android-compose/app/src/main/java/com/ohmz/tday/compose/feature/todos/TodoListScreen.kt",
      "android-compose/app/src/main/java/com/ohmz/tday/compose/feature/completed/CompletedScreen.kt",
      "android-compose/app/src/main/java/com/ohmz/tday/compose/feature/calendar/CalendarScreen.kt",
      "android-compose/app/src/main/java/com/ohmz/tday/compose/feature/scheduledtaskhome/ScheduledTaskHomeScreen.kt",
      "android-compose/app/src/main/java/com/ohmz/tday/compose/feature/car/CarTaskSurfaceScreen.kt",
    ];
    for (const path of androidRows) {
      const source = readRepo(path);
      expect(source).toContain("rememberTaskRowFirstLineAlignment(");
      expect(source).toContain("verticalAlignment = Alignment.Top");
    }
  });

  /**
   * No client may quietly go back to centring.
   *
   * Written as an absence because that is the shape of the regression: the change is
   * one word per row, it reads as harmless in review, and it is invisible on the
   * ~97 % of rows whose title fits a line. `CarTaskSurfaceScreen` is the one file
   * that still carries `Alignment.CenterVertically` on a task row's trailing glyph
   * and argues it in place — that glyph is what the whole card does when pressed,
   * not a mark on the title — so it is exempted by name rather than by silence.
   */
  it.each(iosRows)("%s has no centred task-row HStack left in it", (path) => {
    // The two spellings a task row is opened with. Matching the row's own spacing
    // constant rather than `.center` alone is what keeps this claim about task rows:
    // plenty of other HStacks on these screens are centred and should stay that way.
    expect(readRepo(path)).not.toMatch(
      /HStack\(alignment: \.center, spacing: (TodoTimelineMetrics\.minimalRowContentSpacing|TdayTaskRowMetrics\.contentSpacing)\)/,
    );
  });
});

File: android-compose/app/src/test/java/com/ohmz/tday/compose/core/ui/TaskRowFirstLineAlignmentTest.kt
package com.ohmz.tday.compose.core.ui

import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.unit.Density
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.em
import androidx.compose.ui.unit.sp
import java.io.File
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * The row-stacking rule, asserted where it can be.
 *
 * jsdom lays nothing out and a JVM unit test has no layout pass either, so nothing
 * here can look at a pixel. What it CAN do is check the arithmetic the layout will
 * be given, which is the whole of the fix: the insets are a pure function of a text
 * style's line box and a control's height, and that function is
 * [taskRowFirstLineAlignment].
 *
 * Two kinds of claim, and they are deliberately different in kind.
 *
 * The ARITHMETIC claims pin literals — 12 dp for the task list, 2 dp for Completed,
 * 7 dp for the car surface. Re-deriving the expected value from the same call would
 * assert `f(x) == f(x)` and pass no matter what `f` became, which is the trap
 * `TdayTaskRowSkeletonTest` names when it writes `48.dp` out by hand rather than
 * reading it back out of the metrics.
 *
 * The INVARIANT claims go the other way: they state the property the whole change
 * exists to guarantee — the control's centre and the first line's centre are the
 * same point — and sweep it across font scales and control sizes, because the one
 * thing a hand-written inset cannot do is stay right when the user turns the text
 * up. That is the sweep an emulator would have been for.
 */
class TaskRowFirstLineAlignmentTest {

    private val density = Density(density = 2f, fontScale = 1f)

    /** The shipped `titleMedium`: 18 sp on a 24 sp line. */
    private val titleMedium = TextStyle(fontSize = 18.sp, lineHeight = 24.sp)

    @Test
    fun `the task list's 48 dp toggle drops its title by the 12 dp the row shipped`() {
        // TodoListScreen had `top = TdayDimens.SpacingLg` hand-written under the comment
        // "top pad centres the first title line against the (taller) toggle". That 12 dp
        // is what this call now returns, which is the evidence that the derivation is the
        // existing decision rather than a new one — the one row that was already right
        // does not move by so much as a pixel.
        val alignment = taskRowFirstLineAlignment(density, titleMedium, controlHeight = 48.dp)

        assertEquals(24.dp, alignment.titleLineHeight)
        assertEquals(24.dp, alignment.firstLineCenter)
        assertEquals(12.dp, alignment.titleTopInset)
        // The toggle is the tallest thing in the row, so it is what the line is placed
        // against and it takes no inset of its own.
        assertEquals(0.dp, alignment.topInsetFor(48.dp))
        // The trailing list/priority indicators are 18 dp.
        assertEquals(15.dp, alignment.topInsetFor(18.dp))
    }

    @Test
    fun `Completed's shorter 28 dp toggle asks for a different inset, not the same one`() {
        // The screen in the bug report. Its restore toggle is 28 dp, not 48, so a single
        // constant copied from the task list would have dropped this title 10 dp too far
        // and taken the whole feed's rhythm with it.
        val alignment = taskRowFirstLineAlignment(density, titleMedium, controlHeight = 28.dp)

        assertEquals(14.dp, alignment.firstLineCenter)
        assertEquals(2.dp, alignment.titleTopInset)
        assertEquals(0.dp, alignment.topInsetFor(28.dp))
        assertEquals(5.dp, alignment.topInsetFor(18.dp))
    }

    @Test
    fun `a control SHORTER than the line takes the inset instead of giving it`() {
        // The car surface leads with a 10 dp priority dot. Nothing about the text moves —
        // the dot comes down to meet it. Both directions are the same formula, which is
        // why there is one function and not two.
        val alignment = taskRowFirstLineAlignment(density, titleMedium, controlHeight = 10.dp)

        assertEquals(12.dp, alignment.firstLineCenter)
        assertEquals(0.dp, alignment.titleTopInset)
        assertEquals(7.dp, alignment.topInsetFor(10.dp))
    }

    @Test
    fun `a row whose title line box is overridden is aligned to the override`() {
        // ScheduledTaskHomeScreen writes `lineHeight = 22.sp` at the call site rather than
        // taking titleMedium's 24. A derivation that read the theme would align that row to
        // a line box it is not drawn in — off by 1 dp, forever, invisibly.
        val overridden = titleMedium.copy(lineHeight = 22.sp)
        val alignment = taskRowFirstLineAlignment(density, overridden, controlHeight = 48.dp)

        assertEquals(22.dp, alignment.titleLineHeight)
        assertEquals(13.dp, alignment.titleTopInset)
    }

    @Test
    fun `the inset tracks the font scale, which is the whole reason it is derived`() {
        // A 48 dp control is dp and a 24 sp line is sp, so the gap between them is not a
        // constant — it closes as the user turns the text up and is gone entirely before
        // 3x. The hand-written 12 dp was wrong by 2 dp at 1.5x, by 6 at 2x and by 12 at 3x,
        // in the accessibility setting where being wrong shows most.
        //
        // The numbers are also not `24 * scale`, and that is the point of asserting them
        // rather than computing them: since Android 14, sp above 1x is converted through a
        // NON-LINEAR curve, so 24 sp at 1.5x is 28 dp and not 36. Anybody deriving this by
        // hand would have got the linear answer, and would have been wrong by 4 dp at the
        // one scale that is easiest to reach for by hand.
        val expected = mapOf(
            0.85f to 13.8f,
            1f to 12f,
            1.15f to 11.4f,
            1.3f to 10.8f,
            1.5f to 10f,
            2f to 6f,
            3f to 0f,
        )
        expected.forEach { (scale, inset) ->
            val alignment = taskRowFirstLineAlignment(
                Density(density = 2f, fontScale = scale),
                titleMedium,
                controlHeight = 48.dp,
            )
            assertEquals("at fontScale $scale", inset, alignment.titleTopInset.value, 0.01f)
        }
    }

    @Test
    fun `the control's centre and the first line's centre are always the same point`() {
        // The property the fix exists to guarantee, swept rather than sampled. If this
        // holds, a single-line row cannot have moved either: with one line, the text
        // column's centre IS its first line's centre, so centring the control on one is
        // exactly what `Alignment.CenterVertically` was already doing.
        val controls = listOf(10.dp, 18.dp, 24.dp, 28.dp, 38.dp, 48.dp)
        val scales = listOf(0.85f, 1f, 1.15f, 1.3f, 1.5f, 1.8f, 2f)
        for (control in controls) {
            for (scale in scales) {
                val alignment = taskRowFirstLineAlignment(
                    Density(density = 2f, fontScale = scale),
                    titleMedium,
                    controlHeight = control,
                )
                val controlCenter = alignment.topInsetFor(control) + control / 2
                val firstLineCenter = alignment.titleTopInset + alignment.titleLineHeight / 2
                assertNear(
                    "control ${control.value} dp at fontScale $scale",
                    firstLineCenter,
                    controlCenter,
                )
            }
        }
    }

    @Test
    fun `a trailing indicator lands on the same centre the control does`() {
        // The flag in the screenshot. Deciding it separately is how a half-done fix
        // happens: the toggle moves to line one and the flag stays in the gap.
        val controls = listOf(28.dp, 48.dp)
        val indicators = listOf(13.dp, 14.dp, 18.dp, 24.dp)
        for (control in controls) {
            val alignment = taskRowFirstLineAlignment(density, titleMedium, control)
            for (indicator in indicators) {
                assertNear(
                    "indicator ${indicator.value} dp beside a ${control.value} dp control",
                    alignment.firstLineCenter,
                    alignment.topInsetFor(indicator) + indicator / 2,
                )
            }
        }
    }

    @Test
    fun `nothing is ever asked to hang above the row's top edge`() {
        // `topInsetFor` is the only place a negative padding could have been produced, and
        // Compose throws on one rather than clamping. A 38 dp indicator beside a 24 sp line
        // is not a shape any row draws today — it is the shape the next row might.
        val alignment = taskRowFirstLineAlignment(density, titleMedium, controlHeight = 24.dp)
        assertEquals(0.dp, alignment.topInsetFor(38.dp))
        assertTrue(alignment.topInsetFor(96.dp) >= 0.dp)
    }

    @Test
    fun `a theme that spells its line height in em falls back rather than throwing`() {
        // `TextUnit.toDp()` throws on anything that is not sp, and a row that crashed the
        // feed it is stacking would be a poor trade for exact arithmetic. Same argument,
        // and the same fallback, as `SkeletonTextBar` two functions down.
        val em = TextStyle(fontSize = 18.sp, lineHeight = 1.4.em)
        val alignment = taskRowFirstLineAlignment(density, em, controlHeight = 48.dp)
        assertEquals(24.dp, alignment.titleLineHeight)
        assertEquals(12.dp, alignment.titleTopInset)
    }

    /**
     * Which rows are stacked this way — the structural half of the claim.
     *
     * The arithmetic above is worth nothing if a screen quietly goes back to centring, and
     * "everywhere" was the whole of the request: a fix applied to the screen in the report
     * and to no other leaves the same defect in eight more places. So the rows are pinned
     * by file AND BY COUNT, which is the part a membership test cannot do: `CalendarScreen`
     * draws two of these rows, and a file-level `contains("Alignment.Top")` is satisfied by
     * the first of them no matter what the second does. That is not hypothetical — it is
     * how a half-converted completed row shipped past this test once already.
     */
    @Test
    fun `every task row in the app stacks against its title's first line`() {
        val expected = mapOf(
            "core/ui/TdayTaskRowSkeleton.kt" to 1,
            "feature/todos/TodoListScreen.kt" to 3,
            "feature/completed/CompletedScreen.kt" to 1,
            "feature/calendar/CalendarScreen.kt" to 2,
            "feature/scheduledtaskhome/ScheduledTaskHomeScreen.kt" to 1,
            "feature/car/CarTaskSurfaceScreen.kt" to 1,
        )
        assertEquals(
            "a task row either derives its first line or it does not draw one, and a file " +
                "that gains or loses one says so here. Add the new row, or take the retired " +
                "one out, in the same commit",
            expected,
            rowBlocks().groupingBy { it.label.substringBefore(':') }.eachCount(),
        )
        rowBlocks().forEach { row ->
            assertTrue(
                "${row.label} derives a first line and then stacks its row centred anyway",
                row.text.contains("verticalAlignment = Alignment.Top"),
            )
        }
    }

    /**
     * The leading control is inset too, and this is the half the first cut left out.
     *
     * Only the TEXT was ever given an inset, which is exactly right while the control is
     * the taller of the two and silently wrong the moment it is not: a 28 dp toggle beside
     * a line box that an accessibility font scale has grown to 36 dp is the row's SHORTEST
     * element, and an un-inset one sits 4 dp above the line it is supposed to be the bullet
     * for — above its own trailing flag, too, which did get the call. `topInsetFor` is 0 dp
     * in the common case, so asking for it costs nothing and closes the one direction the
     * row could not otherwise express.
     *
     * Asserted against the control the row DERIVED against, by name: a row that aligns to a
     * 48 dp target and then insets a 24 dp glyph has answered a different question.
     */
    @Test
    fun `every row insets the control it derived its line against`() {
        rowBlocks().forEach { row ->
            val control = CONTROL_HEIGHT.find(row.text)?.groupValues?.get(1)
                ?: error("${row.label} derives a first line without naming a control height")
            assertTrue(
                "${row.label} derives its line against $control and then never puts " +
                    "$control on it — `topInsetFor($control)` is 0 dp at fontScale 1, and " +
                    "is the whole fix above it",
                row.text.contains("topInsetFor($control)"),
            )
        }
    }

    /**
     * Every trailing mark in those rows is on the same line, per MARK and not per file.
     *
     * A trailing mark is found structurally rather than by eye: a `padding(` with an `end`
     * component, on an element whose own call mentions one of the `…TrailingIconSize`
     * tokens. That picks out the list/priority containers and nothing else — a row's own
     * `horizontal`/`vertical` padding has no `end`, and the swipe tray's `end` padding
     * belongs to a call that sizes no mark.
     *
     * The count is pinned for the reason the row count is: a scan that silently matched
     * nothing would pass this test every time.
     */
    @Test
    fun `every trailing mark in those rows is dropped onto the same line`() {
        var checked = 0
        rowBlocks().forEach { row ->
            paddingGroups(row.text).forEach { group ->
                if (!group.text.contains("end =")) return@forEach
                val token = TRAILING_TOKEN.find(enclosingCall(row.text, group.start))
                    ?: return@forEach
                checked++
                assertTrue(
                    "${row.label} stacks from the top and then leaves a ${token.value} mark " +
                        "at the row's top edge. It is an annotation on the title, so it takes " +
                        "`top = firstLine.topInsetFor(${token.value})` like the toggle does",
                    group.text.contains("topInsetFor("),
                )
            }
        }
        assertEquals(
            "the trailing-mark scan found a different number of marks than the tree has. " +
                "Update the count with the row that gained or lost one",
            7,
            checked,
        )
    }

    /** One row's worth of source: its derivation, and everything up to the next row's. */
    private data class RowBlock(val label: String, val text: String)

    /** A `padding(...)` argument list, and where it starts inside its row block. */
    private data class SourceRange(val start: Int, val text: String)

    /**
     * Every row in the tree, one entry each — NOT one entry per file.
     *
     * A row begins where it derives its first line and ends where the next row derives
     * its own, which is what makes the assertions above per-row properties. Only the
     * `val … = remember…` call form opens a block, so the declaration of
     * `rememberTaskRowFirstLineAlignment` itself does not read as a row.
     */
    private fun rowBlocks(): List<RowBlock> = composeSources()
        .filterValues { it.readText().contains("rememberTaskRowFirstLineAlignment(") }
        .flatMap { (path, file) ->
            val text = file.readText()
            val starts = DERIVATION.findAll(text).map { it.range.first }.toList()
            starts.mapIndexed { index, start ->
                RowBlock(
                    label = "$path:${text.take(start).count { it == '\n' } + 1}",
                    text = text.substring(start, starts.getOrNull(index + 1) ?: text.length),
                )
            }
        }

    /** Every `padding(` argument list in [text], balanced rather than line-matched. */
    private fun paddingGroups(text: String): List<SourceRange> {
        val groups = mutableListOf<SourceRange>()
        var from = 0
        while (true) {
            val open = text.indexOf("padding(", from)
            if (open < 0) return groups
            val close = matchForward(text, open + "padding(".length - 1)
            groups += SourceRange(open, text.substring(open, close))
            from = close
        }
    }

    /**
     * The call that owns the argument list [at] sits in — its arguments and, when it has
     * one, its trailing lambda. That is what makes "an element that sizes a trailing mark"
     * a structural question rather than a guess about how far away the mark is written.
     */
    private fun enclosingCall(text: String, at: Int): String {
        var depth = 0
        var index = at - 1
        while (index >= 0) {
            val char = text[index]
            if (char == ')') {
                depth++
            } else if (char == '(') {
                if (depth == 0) break
                depth--
            }
            index--
        }
        if (index < 0) return ""
        val argsEnd = matchForward(text, index)
        var body = argsEnd
        while (body < text.length && text[body].isWhitespace()) body++
        if (body >= text.length || text[body] != '{') return text.substring(index, argsEnd)
        var braces = 1
        var end = body + 1
        while (braces > 0 && end < text.length) {
            when (text[end]) {
                '{' -> braces++
                '}' -> braces--
            }
            end++
        }
        return text.substring(index, end)
    }

    /** The index just past the `)` that closes the `(` at [open]. */
    private fun matchForward(text: String, open: Int): Int {
        var depth = 1
        var index = open + 1
        while (depth > 0 && index < text.length) {
            when (text[index]) {
                '(' -> depth++
                ')' -> depth--
            }
            index++
        }
        return index
    }

    private fun assertNear(message: String, expected: Dp, actual: Dp) {
        assertEquals(message, expected.value, actual.value, 0.001f)
    }

    private companion object {
        /** Only the call form opens a row block — never the function's own declaration. */
        val DERIVATION = Regex("""val\s+\w+\s*=\s*rememberTaskRowFirstLineAlignment\(""")
        val CONTROL_HEIGHT = Regex("""controlHeight\s*=\s*([\w.]+)""")
        val TRAILING_TOKEN = Regex("""\b\w*TrailingIconSize\b""")
    }

    private fun composeSources(): Map<String, File> {
        val root = generateSequence(File(".").canonicalFile) { it.parentFile }
            .map { File(it, "app/src/main/java/com/ohmz/tday/compose") }
            .firstOrNull { it.isDirectory }
            ?: error("could not locate the compose source root from ${File(".").canonicalPath}")
        return root.walkTopDown()
            .filter { it.isFile && it.extension == "kt" }
            .associateBy { it.relativeTo(root).invariantSeparatorsPath }
    }
}
